esphome/components/cse7761/cse7761.cpp

```diff
@@ -0,0 +1,242 @@
+
+static const uint8_t CSE7761_SPECIAL_COMMAND = 0xEA;   // Start special command
+static const uint8_t CSE7761_CMD_RESET = 0x96;         // Reset command, after receiving the command, the chip resets
+static const uint8_t CSE7761_CMD_CLOSE_WRITE = 0xDC;   // Close write operation
+static const uint8_t CSE7761_CMD_ENABLE_WRITE = 0xE5;  // Enable write operation
+
+enum CSE7761 { RMS_IAC, RMS_IBC, RMS_UC, POWER_PAC, POWER_PBC, POWER_SC, ENERGY_AC, ENERGY_BC };
+
+void CSE7761Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up CSE7761...");
+  this->write_(CSE7761_SPECIAL_COMMAND, CSE7761_CMD_RESET);
+  uint16_t syscon = this->read_(0x00, 2);  // Default 0x0A04
+  if ((0x0A04 == syscon) && this->chip_init_()) {
+    this->write_(CSE7761_SPECIAL_COMMAND, CSE7761_CMD_CLOSE_WRITE);
+    ESP_LOGD(TAG, "CSE7761 found");
+    this->data_.ready = true;
+  } else {
+    this->mark_failed();
+  }
+}
+
+void CSE7761Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "CSE7761:");
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with CSE7761 failed!");
+  }
+  LOG_UPDATE_INTERVAL(this);
+  this->check_uart_settings(38400, 1, uart::UART_CONFIG_PARITY_EVEN, 8);
+}
+
+float CSE7761Component::get_setup_priority() const { return setup_priority::DATA; }
+
+void CSE7761Component::update() {
+  if (this->data_.ready) {
+    this->get_data_();
+  }
+}
+
+void CSE7761Component::write_(uint8_t reg, uint16_t data) {
+  uint8_t buffer[5];
+
+  buffer[0] = 0xA5;
+  buffer[1] = reg;
+  uint32_t len = 2;
+  if (data) {
+    if (data < 0xFF) {
+      buffer[2] = data & 0xFF;
+      len = 3;
+    } else {
+      buffer[2] = (data >> 8) & 0xFF;
+      buffer[3] = data & 0xFF;
+      len = 4;
+    }
+    uint8_t crc = 0;
+    for (uint32_t i = 0; i < len; i++) {
+      crc += buffer[i];
+    }
+    buffer[len] = ~crc;
+    len++;
+  }
+
+  this->write_array(buffer, len);
+}
+
+bool CSE7761Component::read_once_(uint8_t reg, uint8_t size, uint32_t *value) {
+  while (this->available()) {
+    this->read();
+  }
+
+  this->write_(reg, 0);
+
+  uint8_t buffer[8] = {0};
+  uint32_t rcvd = 0;
+
+  for (uint32_t i = 0; i <= size; i++) {
+    int value = this->read();
+    if (value > -1 && rcvd < sizeof(buffer) - 1) {
+      buffer[rcvd++] = value;
+    }
+  }
+
+  if (!rcvd) {
+    ESP_LOGD(TAG, "Rx none");
+    return false;
+  }
+
+  rcvd--;
+  uint32_t result = 0;
+  // CRC check
+  uint8_t crc = 0xA5 + reg;
+  for (uint32_t i = 0; i < rcvd; i++) {
+    result = (result << 8) | buffer[i];
+    crc += buffer[i];
+  }
+  crc = ~crc;
+  if (crc != buffer[rcvd]) {
+    return false;
+  }
+
+  *value = result;
+  return true;
+}
+
+uint32_t CSE7761Component::read_(uint8_t reg, uint8_t size) {
+  bool result = false;  // Start loop
+  uint8_t retry = 3;    // Retry up to three times
+  uint32_t value = 0;   // Default no value
+  while (!result && retry) {
+    retry--;
+    result = this->read_once_(reg, size, &value);
+  }
```

Comment:
Maybe a log message would be nice here in the unlikely event that reading fails thrice:
```suggestion
  while (!result && retry > 0) {
    retry--;
    if (this->read_once_(reg, size, &value))
      return value;
  }
  ESP_LOGE(TAG, "Reading register %hhu failed!", reg);
```